attest/ake/src/initiator.rs

```diff
@@ -152,54 +158,94 @@ where
+                    let verifier = DcapVerifier::new(input.identities, input.time, report_data);
+                    let evidence = Evidence::new(quote, collateral)
+                        .map_err(|_| Error::AttestationEvidenceDeserialization)?;
+                    let verification_output = verifier.verify(&evidence);
+                    if verification_output.is_success().into() {
+                        Ok((
+                            Ready {
+                                writer: result.initiator_cipher,
+                                reader: result.responder_cipher,
+                                binding: result.channel_binding,
+                            },
+                            remote_evidence,
+                        ))
+                    } else {
+                        let display_tree =
+                            VerificationTreeDisplay::new(&verifier, verification_output);
+                        Err(Error::AttestationEvidenceVerification(
+                            VerifierError::Verification(display_tree.to_string()),
+                        ))
+                    }
+                } else {
+                    let remote_report = VerificationReport::decode(output.payload.as_slice())
+                        .map_err(|_| Error::AttestationEvidenceDeserialization)?;
+
+                    let identities = input.identities;
+                    let mut verifier = Verifier::default();
+                    verifier.identities(&identities).debug(DEBUG_ENCLAVE);
+
+                    // We are not returning the report data and instead returning the raw report
+                    // since that also includes the signature and certificate chain.
+                    // However, we still make sure the report contains valid data
+                    // before we continue by calling `.verify`. Callers can then
+                    // safely construct a VerificationReportData object out of the
+                    // VerificationReport returned.
+                    let _report_data = verifier
+                        .report_data(
+                            &result
+                                .remote_identity
+                                .ok_or(Error::MissingRemoteIdentity)?
+                                .map_bytes(|bytes| {
+                                    ReportDataMask::try_from(bytes)
+                                        .map_err(|_| Error::BadRemoteIdentity)
+                                })?,
+                        )
+                        .verify(&remote_report)?;
+                    Ok((
+                        Ready {
+                            writer: result.initiator_cipher,
+                            reader: result.responder_cipher,
+                            binding: result.channel_binding,
+                        },
+                        EvidenceKind::Epid(remote_report),
+                    ))
+                }
             }
         }
     }
 }
```

Comment:
This may be a bit heavy of a suggestion, pick and choose as desired
- As currently written it's 80 lines of handling a variant, which feels a bit big.
  - It's using the `EvidenceKind` into/from that my comment probably misled, this is another "legacy" usage location.
  - Subjectively it may be a bit easier to see how each condition/variant differs
- It decodes to `DcapEvidence` to avoid the 3 decodes of each field
- The `remote_identity` comparison in `verify_dcap_evidence`, I'm not sure about, I would like it to be part of the "verify()", but it's probably ok for a one off
- Prevents the duplication of the `Ready`

```suggestion
                let evidence = EvidenceKind::from_bytes(output.payload.as_slice())
                    .map_err(|_| Error::AttestationEvidenceDeserialization)?;
                let remote_identity = result
                    .remote_identity
                    .as_ref()
                    .ok_or(Error::MissingRemoteIdentity)?;
                match &evidence {
                    EvidenceKind::Dcap(prost_evidence) => {
                        let dcap_evidence = DcapEvidence::try_from(prost_evidence)
                            .map_err(|_| Error::AttestationEvidenceDeserialization)?;
                        verify_dcap_evidence(
                            dcap_evidence,
                            &input.identities,
                            input.time,
                            remote_identity,
                        )?;
                    }
                    EvidenceKind::Epid(report) => {
                        verify_verification_report(report, &input.identities, remote_identity)?;
                    }
                }
                Ok((
                    Ready {
                        writer: result.initiator_cipher,
                        reader: result.responder_cipher,
                        binding: result.channel_binding,
                    },
                    evidence,
                ))
            }
        }
    }
}

fn verify_verification_report<PubKey: KexPublic>(
    report: &VerificationReport,
    identities: &Vec<TrustedIdentity>,
    remote_identity: &PubKey,
) -> Result<(), Error> {
    let mut verifier = Verifier::default();
    verifier.identities(identities).debug(DEBUG_ENCLAVE);

    // We are not returning the report data and instead returning the raw report
    // since that also includes the signature and certificate chain.
    // However, we still make sure the report contains valid data
    // before we continue by calling `.verify`. Callers can then
    // safely construct a VerificationReportData object out of the
    // VerificationReport returned.
    let _report_data = verifier
        .report_data(&remote_identity.map_bytes(|bytes| {
            ReportDataMask::try_from(bytes).map_err(|_| Error::BadRemoteIdentity)
        })?)
        .verify(report)?;
    Ok(())
}

fn verify_dcap_evidence<PubKey: KexPublic>(
    dcap_evidence: DcapEvidence,
    identities: &Vec<TrustedIdentity>,
    time: Option<DateTime>,
    remote_identity: &PubKey,
) -> Result<(), Error> {
    let DcapEvidence {
        quote,
        collateral,
        report_data,
    } = dcap_evidence;
    if remote_identity.map_bytes(|bytes| bytes != report_data.key().to_bytes().as_slice()) {
        return Err(Error::BadRemoteIdentity);
    }

    let verifier = DcapVerifier::new(identities, time, report_data);
    let evidence =
        Evidence::new(quote, collateral).map_err(|_| Error::AttestationEvidenceDeserialization)?;
    let verification = verifier.verify(&evidence);
    if verification.is_failure().into() {
        let display_tree = VerificationTreeDisplay::new(&verifier, verification);
        Err(VerifierError::Verification(display_tree.to_string()).into())
    } else {
        Ok(())
    }
}
```

Reply:
I like this idea, but it needs a bit of work to fit with the changes decoding EvidenceKind instead of DcapEvidence. I've implemented something similar.